C ABI Changes for `wasm32-unknown-unknown`

alexcrichton opened the pull request:
This adds a post for changes discussed in
https://github.com/rust-lang/rust/issues/122532 and https://github.com/rust-lang/rust/pull/138601, notably as a result of this decision:
https://github.com/rust-lang/rust/issues/122532#issuecomment-2740974434



[Rendered](https://github.com/alexcrichton/blog.rust-lang.org/blob/wasm-c-abi-changes/content/c-abi-changes-for-wasm32-unknown-unknown.md)

alexcrichton commented:
I'll note that at this time I'm opening this up to share with folks and get feedback, there's no concrete timeline for when to publish this. My hope for now is "soon after https://github.com/rust-lang/rust/pull/138601"

alexcrichton commented:
I'll also note, wow, this issue has a huge and sprawling history. I've surely missed something and wanted to clarify that it's not intentional. Let me know!

Manishearth commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,258 @@
+The time has now come to correct this historical mistake and the Rust compiler
+will soon be using a correct ABI definition for the `wasm32-unknown-unknown`
+target. This means, however, that generated WebAssembly binaries will be
+different than before?
```
nit: question mark is probably a typo?

Manishearth commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,258 @@
+and the WebAssembly type system. This includes details such as in-memory layouts
+and translations of a C function signature to a WebAssembly function signature.
+
+## How is `wasm32-unknown-unknown` non-standard?
```
suggestion: if you'd like to link to a document describing this in its full extent (mostly), you can link to https://github.com/rust-diplomat/diplomat/blob/main/docs/wasm_abi_quirks.md

alexcrichton replied:
I would very much like to link that, thank you!

Manishearth commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,258 @@
+with `-Zwasm-c-abi=spec`. If all that passes then you're good to go and the
+upcoming change to the C ABI will not affect your project.
+
+## I'm affected, now what?
```
this looks great, thanks!

ehuss commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,258 @@
+
+When the `wasm32-unknown-unknown` target [was originally added][inception] in
+2017 not much care was given to the exact definition of the `extern "C"` ABI at
+the time. In 2018 [an ABI definition wasm added just for wasm][orig-abi] and the
```
```suggestion
the time. In 2018 [an ABI definition was added just for wasm][orig-abi] and the
```

ehuss commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,263 @@
+
+Notably you can see here that the struct `Pair` was "splatted" into its two
+components so the actual `$pair_add` function takes two arguments, the `x` and
+`y` fields. The [tool-conventions], however [specifically says] that "other
```
`[specifically says]` doesn't link to anything.

alexcrichton replied:
Oops thanks, I was originally going to link something here but realized that I couldn't get a more specific link so I inlined the words instead and forgot to remove the brackets!

RalfJung reviewed (commented):
That's a lot of detail, thanks!

I think what would help is to have a short sentence, before the history lessons, telling people where to jump to in case they just want to get the gist of it and read up on the history later.

bjorn3 commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,263 @@
+
+At the time this was discovered it was found to be significantly difficult to
+fix `wasm-bindgen` to not rely on this splatting behavior. At the time it also
+wasn't though to be a widespread issue nor was it costly for the compiler to
```
*thought?

alexcrichton commented:
I believe I've addressed all feedback here and I've also makred this as ready-to-go, but I'll drop a comment here after a nightly is available since I think it would be best to wait for a compiler first before publishing this.

alexcrichton commented:
https://github.com/rust-lang/rust/pull/138601 has now landed, but it didn't make today's nightly, so best to wait one more day for this

ehuss commented:
Just to ensure this doesn't block without a clear owner, I assume this is on behalf of the compiler team, so @davidtwco or @wesleywiser I assume this would be yours to approve?

ehuss commented on posts/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,273 @@
++++
+layout = "post"
+date = 2024-03-20
```
And make sure it is the correct year. 😜

alexcrichton replied:
Oop yes thanks! Updated to tomorrow, but happy to update this again if needed

senekor commented:
Note that posts were moved to the `content/` directory, sorry for the churn 🙂

alexcrichton commented:
Updated 👍

lcnr reviewed (commented):
please update the date to today, after that r=me

lcnr commented on content/c-abi-changes-for-wasm32-unknown-unknown.md:
```diff
@@ -0,0 +1,273 @@
++++
+layout = "post"
+date = 2025-04-03
```
```suggestion
date = 2025-04-04
```
